wgsl/index.bs

```diff
@@ -6981,13 +6979,326 @@ See [[#statements]] and [[#function-calls]].
+
+Most built-in functions have tags of:
+- for the control-flow in which they can be called: MustBeUniformForControlFlowAfter
+- for the function as a whole: NoRestriction
+- for each parameter: MustBeUniformForReturnValue
+
+Here is the list of exceptions:
+- All functions in [[#derivative-builtin-functions]] are AlwaysMustBeUniform and ReturnValueCannotBeUniform
+- All functions in [[#texturesample]], [[#texturesamplebias]], [[#texturesamplecompare]], and [[#sync-builtin-functions]] are AlwaysMustBeUniform
+- TODO: what about [[#texturesamplelevel]] and [[#texturestore]] ?
+
+### Uniformity rules for expressions ### {#uniformity-expressions}
+
+The rules for analyzing expressions take as argument both the expression itself and the node corresponding to control flow at the beginning of it (which we'll note "CF" below) and return the following:
+
+* A node corresponding to control flow at the exit of it
+* A node corresponding to its value
+* A set of new nodes and edges to add to the graph
+
+Like for statements, if the set of behaviors for an expression is {Next}, then the node corresponding to control flow at the exit of it is replaced by *CF* (the control flow at the beginning of it), overriding what the rules below say about the resulting control-flow node.
+
+<table class='data'>
+  <caption>Uniformity rules for expressions (in normal rvalue position)</caption>
+  <thead>
+    <tr><th>Expression<th>New nodes<th>Recursive analyses<th>Resulting control flow node (if behavior is not {Next}), value node<th>New edges
+  </thead>
+  <tr><td class="nowrap">e1 || e2
+      <td>
+      <td class="nowrap">(*CF*, *e1*) => (*CF1*, *V1*)<br>
+          (*V1*, *e2*) => (*CF2*, *V2*)
+      <td class="nowrap">*CF2*, *V2*
+      <td>
+  <tr><td class="nowrap">e1 && e2
+      <td>
+      <td class="nowrap">(*CF*, *e1*) => (*CF1*, *V1*)<br>
+          (*V1*, *e2*) => (*CF2*, *V2*)
+      <td class="nowrap">*CF2*, *V2*
+      <td>
+  <tr><td class="nowrap">Literal
+      <td>
+      <td>
+      <td class="nowrap">*CF*, *CF*
+      <td>
+   <tr><td class="nowrap">reference to local variable "x"
+      <td>*Result*
+      <td class="nowrap">*X* is the node corresponding to "x"
+      <td class="nowrap">*CF*, *Result*
+      <td class="nowrap">*Result* -> {*CF*, *X*}
+   <tr><td class="nowrap">reference to uniform built-in variable "x"
+      <td>
+      <td>
+      <td class="nowrap">*CF*, *CF*
+      <td>
+   <tr><td class="nowrap">reference to non-uniform built-in variable "x"
+      <td>
+      <td>
+      <td class="nowrap">*CF*, *CannotBeUniform*
+      <td>
+   <tr><td class="nowrap">reference to global variable "x"
+      <td>
+      <td>
+      <td class="nowrap">*CF*, *CannotBeUniform*
+      <td>
```

Comment:
Read-only variables shouldn't be non-uniform automatically. It should only depend on uniform access.

Reply:
Good idea, I split that case into read-only / not read-only.